### PR TITLE
Transformer backbones: changes to CLI args and input / output upsampling for Swin 

### PR DESCRIPTION
**Pull request by dmizr**
- Changes the CLI arguments for Swin and XCiT to be more similar to those of ResNet / ShuffleNet, and adds 
- Optional input upsampling for Swin
- FPN for output feature map upsampling for Swin
- Stochastic depth rate for Swin can be changed through CLI

### CODE REVIEW

**Comment by svenkreiss** [comment]
Thanks! In principle, I like these changes. 

Be careful with depending on class variables in the forward pass. After training, the checkpoint serializes (pickles) only the instance variables. To have the same behavior during training and prediction without having to know which cli args were used, maybe introduce instance variables in the constructor that copy the class variables.

**Comment by dmizr** [comment]
> Be careful with depending on class variables in the forward pass. After training, the checkpoint serializes (pickles) only the instance variables. To have the same behavior during training and prediction without having to know which cli args were used, maybe introduce instance variables in the constructor that copy the class variables.

Thanks, that's a really great point. As a fix, I opted to create both `in_block` and `out_block` modules for Swin. I decided to make the out block its own class (`FPNOutBlock`), which has the added benefit of making it usable by any network which outputs multi-scale feature maps, which is the case for many [mmdet](https://github.com/open-mmlab/mmdetection) backbones using Feature Pyramid Networks. This module could then be further improved later on (e.g. to output higher res feature maps) if needed.

**Comment by dmizr** [comment]
After reconsideration, I opted to add a complete [FPN](https://arxiv.org/abs/1612.03144), although modified to only refine and output the feature map of the chosen level, as openpifpaf does'nt seem to support multi-scale feature maps. This FPN network is usable by any architecture which outputs multi-scale feature maps, and can also perhaps serve as a replacement for PixelShuffle upsampling for these architectures.

File: openpifpaf/network/basenetworks.py [hunk 1]
```diff
@@ -498,102 +498,160 @@ def configure(cls, args: argparse.Namespace):
         cls.pretrained = args.squeezenet_pretrained
 
 
+class FPN(torch.nn.Module):
```
Comment (svenkreiss) [inline openpifpaf/network/basenetworks.py:501]: OpenPifPaf already depends on `torchvision` which does have a FeaturePyramidNetwork implementation. Could this be used here? If not, add a small comment regarding this in your docstring. You had it in the PR comment, but please also add the link to the paper here (just to clarify that you are implementing the original FPN).

**Review by svenkreiss — COMMENTED** [review]
Thanks @dmizr . Overall looks great. Minor comment below.

File: openpifpaf/network/basenetworks.py (hunk 1 above)
Comment (dmizr) [reply openpifpaf/network/basenetworks.py:501]: [torchvision's FPN](https://github.com/pytorch/vision/blob/master/torchvision/ops/feature_pyramid_network.py) refines and outputs all feature maps, which means applying a 3x3 conv on all of the feature maps, even if they don't get used.

Here, this implementation only refines and outputs the desired level, which means that if, for example, I want the feature map of stride 16 from the 3rd FPN level, I won't need to also compute the feature maps of stride 4, 8 and 32.

While it should be possible to avoid refining higher res feature maps with torchvision's FPN by adapting the codebase here, I don't see a way to avoid refining the lower res feature maps. The changes to avoid refining higher res feature maps would also complexify the `__init__` and `forward` of the `SwinTransformer` class, but I think that's minor compared to the impact on compute. 

I updated the docstring and added a link to the paper, but do let me know if you think it's worth switching to torchvision's FeaturePyramidNetwork @svenkreiss .

Comment (svenkreiss) [reply openpifpaf/network/basenetworks.py:504]: That all makes sense to me. Thanks for clarifying! Do you mind putting the first two paragraphs (or something similar) into the docstring? I am pretty sure someone (and it might be me) will have the same question again in six months and the PR discussions here are usually lost.

**Comment by svenkreiss** [comment]
Perfect! Many thanks for the contribution and the detailed clarifications.
